### PR TITLE
DIP1021: Argument Ownership and Function Calls

### PR DESCRIPTION
**Pull request by WalterBright**
Initial implementation of 
https://github.com/dlang/DIPs/blob/793f83911fdc8c88c6ef34e6a36b5e11e3e574e5/DIPs/DIP1021.md

### CODE REVIEW

File: src/dmd/escape.d
Comment (Geod24) [inline src/dmd/escape.d]: So this is... `O((n^2)/2 * (v+r))` in the normal case (no error) (`n == escapeBy.length`, `v == byValue.length`, `r == byRef.length`). That doesn't sound quite right.

Comment (WalterBright) [reply src/dmd/escape.d]: It's bad if there are a lot of parameters, but those cases are awfully rare.

**Comment by look-at-me** [comment]
Don't think this should be added until the DIP actually defines what this is suppose to do. The current DIP is less than bare bones.

File: src/dmd/escape.d
Comment (Geod24) [reply src/dmd/escape.d]: Since there is no usage of this yet, we don't know how impactful it will be. But what we already know is that it doesn't scale at all.

**Review by wilzbach — APPROVED** [review]
It's okay to merge this under a preview flag. It hurts no one and gives people a real chance to play with it while evaluating the DIP. Same reasoning would also apply to other PRs for DIPs.

File: src/dmd/escape.d
Comment (Geod24) [inline src/dmd/escape.d]: This block is not covered

File: src/dmd/escape.d
Comment (Geod24) [inline src/dmd/escape.d]: This block is not covered

**Comment by Geod24** [comment]
Some lines are not covered, e.g. the interaction with `inout` is written but not tested

**Comment by WalterBright** [comment]
Coverage is much better now.

File: src/dmd/globals.d
Comment (JinShil) [inline src/dmd/globals.d]: That URL does not exist.

File: src/dmd/globals.h
Comment (JinShil) [inline src/dmd/globals.h]: That page does not exist on the wiki.

**Comment by Geod24** [comment]
Shall we just give up on review if you are just going to merge right after pushing? I can understand merging something that has been stalled for a while, or that has been approved and rebased without significant change, but you've been working on this for the past few hours, did 7 force pushes, and just added the automerge label without any reviews.

May I suggest using the "72 hours" label if you feel the PR has waited long enough?

**Comment by WalterBright** [comment]
I did fix one minor oversight in the code (delegates and function pointers were not considered). The rest of the force pushes were working on the test cases to get coverage. As for the rest, it has sat around for a month with no comments. If anyone wanted to make a comment on it, they would have by now.

File: src/dmd/globals.h
Comment (WalterBright) [reply src/dmd/globals.h]: It will.

File: src/dmd/cli.d
Comment (JinShil) [inline src/dmd/cli.d]: This URL also appears to be wrong.  I think you mean https://github.com/dlang/DIPs/blob/master/DIPs/DIP1021.md

File: src/dmd/globals.h
Comment (JinShil) [reply src/dmd/globals.h]: I believe you meant this URL: https://github.com/dlang/DIPs/blob/master/DIPs/DIP1021.md

**Review by Geod24 — APPROVED** [review]
Just one minor comment (and the fact that all URL to documentation is broken). Otherwise I guess we can take this out for a spin. Did you check what effect it had on Druntime / Phobos ?

**Comment by WalterBright** [comment]
Fixed URLs.

**Pull request merged**
